common/quotas/ownership_scaled_quota.go

```diff
@@ -0,0 +1,161 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+type (
+	// OwnershipScaledQuota is a RateBurst implementation that scales the RPS and burst quotas linearly with the
+	// fraction of the total shards in the cluster owned by this host. The effective Rate and Burst are both multiplied
+	// by:
+	//
+	//	shardCount / totalShards * constantScaleFactor
+	//
+	// Note that there is no scaling, not even by the constantScaleFactor, until the first shard count update is
+	// received.
+	OwnershipScaledQuota struct {
```

Comment:
nit: given that it implements RateBurst interface, should it be called `OwnershipScaledRateBurst`?

And I guess this implementation will be used to construct a dynamic rate limiter so that it becomes ownership aware?

Reply:
> nit: given that it implements RateBurst interface, should it be called OwnershipScaledRateBurst?

Renamed

> And I guess this implementation will be used to construct a dynamic rate limiter so that it becomes ownership aware?

Yes, this is the most composable way of doing it.